### ISSUE TITLE
`unreachable_from_main` flags angular component classes

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#3733]
It's not super uncommon for an Angular test to declare a component class in a test file (with a main function) which is not referenced anywhere in the file. Currently `unreachable_from_main` flags such a class.

**To Reproduce**

```dart
import 'package:test/test.dart';
import 'package:angular/angular.dart';
import 'package:angular/src/bootstrap/run.dart';

import 'hello_test.template.dart' as ng;

void main() {
  test('bootstrap should create a new element if missing from the DOM', () {
    final injector = appInjector((parent) => parent);
    final ApplicationRef appRef = injector.provideType(ApplicationRef);
    final comp = appRef.bootstrap(ng.createHelloComponentFactory());
    expect(comp.location.text, 'Hello World');
  });
}

@Component(
  selector: 'hello-component',
  template: 'Hello World',
)
class HelloComponent {}
```

**Expected behavior**

A good question. Perhaps `unreachable_from_main` needs to just back off on top-level elements with an annotation. This would solve the grinder issue as well.

Perhaps just annotations from _outside_ the meta library.

**Comment by a14n** [comment]
Do you mean that any top level member with an annotation should be treated as an entry-point ? This seams acceptable.

**Comment by srawlins** [comment]
Hmm, I hadn't meant that, but actually that makes for a very simple impl (whimsical rhyming intended). I think that's a good way to do it.

Oh, but maybe, unless this is too specific and weird, only annotations not from package:meta.

**Comment by a14n** [comment]
What annotation in meta could be problematic? `@immutable` ?

**Comment by srawlins** [comment]
I guess there aren't too many that can apply to top-level elements, but if anything is annotated `@alwaysThrows` or `@experimental` or `@internal`, etc. should not be given a free pass; if something is not reachable but has one of those annotations, it should be flagged.

Same with `@deprecated`, but that's from dart:core I guess.
